rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,957 @@
+    /// If a new transaction is created, push that into state.transactions
+    // TODO future, improved version may be based on current message type and dir, too
+    fn find_or_create_tx(&mut self) -> &mut PgsqlTransaction {
+        // First, check if we should create a new tx (in case the other was completed or there's no tx yet)
+        if  self.state_progress == PgsqlStateProgress::IdleState ||
+            self.state_progress == PgsqlStateProgress::StartupMessageReceived ||
+            self.state_progress == PgsqlStateProgress::ConnectionCompleted ||
+            self.state_progress == PgsqlStateProgress::ReadyForQueryReceived ||
+            self.state_progress == PgsqlStateProgress::SslRejectedReceived {
+                let tx = self.new_tx();
+                self.transactions.push(tx);
+            }
+            // If we don't have to create a new transaction, just return the current one
+            return self.transactions.last_mut().unwrap();
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty requests.
+        if input.len() == 0 {
+            return AppLayerResult::ok();
+        }
+
+        // If there was gap, check we can sync up again.
+        if self.request_gap {
+            // TODO I have changed probe to just return a boolean. will this be an issue, later on?
+            if !probe_ts(input) {
+                // The parser now needs to decide what to do as we are not in sync.
+                // For now, we'll just try again next time.
+                SCLogDebug!("Suricata interprets there's a gap in the request");
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.request_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+
+            // TODO rewrite this, in light of PgsqlStateProgress?
+            // if PgsqlStateProgress is AuthenticationGSS, parse GSS response -> TODO decide if we should offer support for it in the first version
+            // if PgsqlStateProgress is AuthenticationSSPI, parse SSPI response -> make sure I have the proper nom parser
+            SCLogDebug!("In 'parse_request' State Progress is: {:?}", &self.state_progress);
+            match self.state_progress {
+                PgsqlStateProgress::IdleState |
+                PgsqlStateProgress::ConnectionCompleted |
+                PgsqlStateProgress::SslRejectedReceived |
+                PgsqlStateProgress::SslAcceptedReceived |
+                PgsqlStateProgress::ReadyForQueryReceived |
+                PgsqlStateProgress::CommandCompletedReceived => {
+                    match parser::parse_request(start) {
+                        // TODO Question should I also add "&& self.transactions.is_empty()"
+                        Ok((rem, request)) => {
```

Comment:
think it would be good to move this block into a util func to keep the main loop more readable